polyfill/index.d.ts

```diff
@@ -0,0 +1,385 @@
+export namespace Temporal {
+  export type ComparisonResult = -1 | 0 | 1;
+
+  // TODO: the `XXXOptions` types below helper types below are useful while
+  // Temporal is being designed, beacuse they make it easier to understand how
+  // options are shared across types and methods. But once the API is finalized
+  // it may make sense to explicitly list all options inside each method
+  // declaration, because later changes to options accepted by some but not all
+  // methods won't result in a breaking change.
+
+  /**
+   * Options for assigning fields using `with()` or entire objects with
+   * `from()`.
+   * */
+  export type AssignmentOptions = {
+    /**
+     * How to deal with out-of-range values
+     *
+     * - In `'constrain'` mode, out-of-range values are clamped to the nearest
+     *   in-range value.
+     * - In `'balance mode'`, out-of-range values are resolved by balancing them
+     *   with the next highest unit.
+     * - In `'reject mode'`, out-of-range values will cause the function to
+     *   throw a RangeError.
+     *
+     * The default is `'constrain'`.
+     */
+    disambiguation: 'constrain' | 'balance' | 'reject';
+    // TODO: should change name from `disambiguation` to `overflow`? See #607.
```

Comment:
I'd suggest removing the TODO notes and instead just have the types file always reflect the current state of the polyfill. (it would be nice if we could fail CI when they get out of sync, but I'm not sure that's possible?)

Reply:
When is the polyfill going out to npm?  It definitely makes sense to remove these comments before publishing to npm.

In the meantime I was assuming that there's value in highlighting which APIs might be changing-- especially those that seem to have initial consensus but which others might object to. I wanted to provide a way to encourage feedback before changes are finalized as opposed to after. Also to prevent duplicate issues from being filed if changes are already being discussed. But I'm also happy to take them out, let me know what you think.

> (it would be nice if we could fail CI when they get out of sync, but I'm not sure that's possible?)

Yep, this only works if there's other code to get out of sync with.  Because the polyfill implementation and tests are untyped JS, it's harder. The TS compiler can type-check plain JS code and find some problems but it's often hit-or-miss.

One way to solve this would be to be able to run at least some test code through the TS compiler-- essentially to be able to write tests in TS or JS. With at least one TS test for every API schema change, we'd probably be able to catch most out-of-date types.

One thing I could also try would be running some of the tests through the TypeScript compiler-- not to actually build tests in TS, but to trigger errors if test code is doing things that TS can flag as errors. 

The CI part is pretty easy: just include a single run of the TS compiler as part of CI build/test scripts. Having the actual content to compile is the hard part. 

One no-brainer thing to do would be to run the .d.ts file itself through the TS compiler. This would at least catch syntax errors or inconsistencies inside the type definitions themselves, which is better than nothing.

Want me to open an issue for this?

Reply:
Sounds like some great ideas. I think it'd be useful to first add validation of the .d.ts file itself, then we can see how difficult it would be to compile the test files with the TS compiler.

Reply:
@ptomato - circling back on this issue... still looking for your reply:

> When is the polyfill going out to npm? It definitely makes sense to remove these comments before publishing to npm.

> In the meantime I was assuming that there's value in highlighting which APIs might be changing-- especially those that seem to have initial consensus but which others might object to. I wanted to provide a way to encourage feedback before changes are finalized as opposed to after. Also to prevent duplicate issues from being filed if changes are already being discussed. But I'm also happy to take them out, let me know what you think.

Reply:
Oh, sorry about that! Yes, we intended to put the polyfill on npm as soon as calendars and Temporal.parse have landed (basically everything in the https://github.com/tc39/proposal-temporal/milestone/4 milestone). Ideally before the TC39 meeting next week.

Personally I'd rather have no TODOs in the types file but instead make it clear that we encourage feedback on all of the API, not just specific parts, and that changes will only be finalized when the proposal advances to Stage 3.